### ISSUE TITLE
Request for feedback on adding support for sorting and filtering completion items

### ISSUE THREAD

**Issue by thomasjo** [atom/atom-languageclient#133]
This issue is intended as an open discussion around how we might go about adding support for sorting and filtering the "raw" `CompletionItem` results returned by [`textDocument/complete`](https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#completion-request).

## Background
The specific case that spawned the need for this functionality is my work on [ide-cpp](https://github.com/thomasjo/atom-ide-cpp), which is built around the Clangd tool in the Clang/LLVM toolkit. Clangd returns completion items in a semi-random, non-deterministic order, and therefore the results need to be sorted (and filtered) before we converting to AC+ format and present the final results in the user.

Without adding some type of extensibility point to the current API, implementing sorting and filtering requires arguably too much code/effort; the easiest way right now (seemingly) requires creating an adapter that extends (or composes) `AutocompleteAdapter` in order to override the `getSuggestions` method. This is not a huge problem, but also not ideal, hence my proposal for adding some type of extensibility point. However, I am not entirely sure how to implement this in the best and most optimal manner, although I have a few ideas that I hope can serve as a starting point for the discussion. I will start by summarizing my current ideas, and provide further implementation details below.

## Ideas
1. Add instance methods to `AutocompleteAdapter` that allows injecting sort and filter functions, which will be used in `AutocompleteAdapter.getSuggestions`.
2. Add _optional_ instance methods to `AtomLanguageClient` that allows defining sort and filter implementations, which will be used in `AtomLanguageClient.getSuggestions`.
3. Expand sorting in AC+ by implementing support for `sortText` as defined in [`CompletionItem`](https://github.com/Microsoft/language-server-protocol/blob/master/protocol.md#completion-request).

### Idea 1
Extending `AutocompleteAdapter` with support for injecting sort and filter functionality seems like the approach that requires the least amount of effort, while still ticking all the boxes.

This approach will require some minor changes beyond simply adding injection methods, and using the injected bits; right now `AtomLanguageClient` uses lazy-initialization of `AutocompleteAdapter` in `getSuggestions`. In order to inject the implementations into an `AutocompleteAdapter` instance, we might consider either

1. Add an instance method for getting an instance of `AutocompleteAdapter`, which could also perform lazy-initialization.
2. Add a (get) property for getting an instance of `AutocompleteAdapter`, which could also perform lazy-initialization.
3. Eagerly initialize an `AutocompleteAdapter` instance during the construction of `AtomLanguageClient`, and expose this as an instance variable.

I am leaning towards either 2 or 3, with the latter requiring the least amount of effort. Implementing either is trivial though — let us briefly consider the following unpolished implementations of all three approaches:

#### Approach 1 (instance method)
``` js
class AutoLanguageClient {
  autocompleteAdapter () {
    if (!this.autoComplete) this.autoComplete = new AutocompleteAdapter()
    return this.autoComplete
  }
}

class CppLanguageClient extends AutoLanguageClient {
  constructor () {
    super()
    this.autocompleteAdapter().sortItems = function (items) { /* ... */ }
  }
}
```

#### Approach 2 (property)
``` js
class AutoLanguageClient {
  get autocompleteAdapter () {
    delete this.autocompleteAdapter
    return this.autocompleteAdapter = new AutocompleteAdapter()
  }
}

class CppLanguageClient extends AutoLanguageClient {
  constructor () {
    super()
    this.autocompleteAdapter.sortItems = function (items) { /* ... */ }
  }
}
```

#### Approach 3 (constructor initialization)
``` js
class AutoLanguageClient {
  constructor () {
    this.autocompleteAdapter = new AutocompleteAdapter()
  }
}

class CppLanguageClient extends AutoLanguageClient {
  constructor () {
    super()
    this.autocompleteAdapter.sortItems = function (items) { /* ... */ }
  }
}
```

### Idea 2
Instead of adding extensibility points to `AutocompleteAdapter`, add them to `AtomLanguageClient` instead. Although, due to the way `AutocompleteAdapter` is implemented, and how `AtomLanguageClient` uses it, this approach will likely require more effort.

One of the problems I discovered while researching and writing this, is that right now `AutocompleteAdapter` violates the _Single-Responsibility Principle_ since it does more than "adapting" between the two domains. Let us consider the current implementation of `AutocompleteAdapter.getSuggestions`

``` js
async getSuggestions(connection, request) {
  const completionItems = await connection.completion(
    AutocompleteAdapter.requestToTextDocumentPositionParams(request))
  return AutocompleteAdapter.completionItemsToSuggestions(completionItems, request)
}
```

As we can see, it communicates with language servers, and I think maybe this is one of the reasons why I was struggling with finding a clean proposal for how to implement sorting and filtering. W should consider refactoring if we decide to add the proposed extensibility points to `AtomLanguageClient`.

Personally, I feel like this idea is the approach we should take. Filtering and sorting results from a language server is a concern that belongs with the associated language client, not the very generic `AutocompleteAdapter`. _(Note that this "violation" is also present in other adapters.)_

Putting aside the "problems", the essence of the approach in this idea is to add sort and filtering support to `AtomLanguageClient` to make it easy for implementors to provide the functionality. Perhaps something like the following

``` js
class AtomLanguageClient {
  async sortCompletionItems (completionItems) {
    return completionItems
  }

  async getSuggestions(connection, request) {
    const completionItems = await connection.completion(
      AutocompleteAdapter.requestToTextDocumentPositionParams(request))
    const sortedItems = await sortCompletionItems(completionItems)
    return AutocompleteAdapter.completionItemsToSuggestions(sortedItems, request)
  }
}

class CppLanguageClient extends AutoLanguageClient {
  async sortCompletionItems (completionItems) {
    return _.sortBy(completionItems, ['sortText'])
  }
}
```

### Idea 3
Implement support for the optional `sortText` field in `CompletionItem` directly in AC+. This would mean that everyone would get sorting (and filtering) for free whenever provided by a language server. This approach would most likely result in a massive effort up-front, but would save everyone time in the long run.

## Other thoughts, questions, and concerns.
Irregardless of what approach we decide to go for in the end, there are still a few outstanding things I think need to be addressed.

### Order of operations
Assuming both filtering and sorting extensibility points;
1. Sort first, then filter?
2. Filter first, then sort?
3. Combine into one operation, thus leaving the order of operations up to the implementor?

---
This is by no means an exhaustive or complete view of all possibilities, solutions, problems, etc. It is simply the ramblings of a madman. Therefore it might not even be completely coherent, nor logical.

**Comment by damieng** [comment]
I've checked something in to address the sorting - I'm not sure this needs much attention as the order of filter/sort should make no difference and it's hard to see how people would need extensibility for that given it's just a fixed field in the spec with a clear fallback. Perhaps we should reduce the scope of this RFF to just filtering and how that can be made extensible.

Triggering and caching as well as filtering based on the characters entered has had some attention in an open PR at https://github.com/atom/atom-languageclient/pull/128 and a custom conversion of items PR at https://github.com/atom/atom-languageclient/pull/137

**Comment by damieng** [comment]
Conversion hook is now also checked in as well as sorting the items in the array based on the sortText.  

Let me know how you get on and if you need anything further!

**Comment by damieng** [comment]
Fixed in a6a620694b49fe1c2cd51a7d9b698c505ea08416
